File: NoTomorrow/Features/Workout/WorkoutStarter.swift
import Foundation
import SwiftData

/// Builds `Workout` graphs (exercises + set rows) from a routine or from nothing, and hands them to the session.
/// The one start path for Train and Today, so both build the same workout.
@MainActor
enum WorkoutStarter {

    /// What a Start tap asks for.
    enum Request {
        case routine(Routine)
        case empty
    }

    /// Whether a Start tap may go ahead. Only one workout runs at a time: while one is in progress the caller
    /// offers Resume, and "Discard it and start new" only when that workout has no completed sets.
    enum Gate {
        case clear
        case blocked(active: Workout, canDiscard: Bool)
    }

    /// A Start blocked by the workout in progress, while the user decides (`WorkoutStartConflictDialog`).
    struct Conflict: Identifiable {
        let id = UUID()
        let active: Workout
        /// Read when the Start was tapped, so the dialog never reads a row a discard is deleting.
        let activeName: String
        let canDiscard: Bool
        let request: Request

        init(active: Workout, canDiscard: Bool, request: Request) {
            self.active = active
            self.activeName = active.name
            self.canDiscard = canDiscard
            self.request = request
        }
    }

    /// Starts right away, or leaves the conflict on the session for the tab shell to ask about.
    static func requestStart(_ request: Request, in context: ModelContext, session: WorkoutSessionController) {
        switch gate(in: context, session: session) {
        case .clear:
            start(request, in: context, session: session)
        case .blocked(let active, let canDiscard):
            session.startConflict = Conflict(active: active, canDiscard: canDiscard, request: request)
        }
    }

    /// "Discard it and start new" in the conflict dialog. The dialog is part of the tab shell, not a presented alert,
    /// so this runs at the tap: the old workout and its mini bar go at once and the new workout's cover comes up
    /// straight away. With completed sets in the old workout it refuses and brings that one back instead.
    /// The caller skips the rest timer first.
    @discardableResult
    static func resolveByDiscarding(_ conflict: Conflict, in context: ModelContext,
                                    session: WorkoutSessionController) -> Workout {
        session.startConflict = nil
        guard discardForNewStart(conflict.active, in: context, session: session) else { return conflict.active }
        session.hideMiniBarWhileSwapping()
        return start(conflict.request, in: context, session: session)
    }

    static func gate(in context: ModelContext, session: WorkoutSessionController) -> Gate {
        guard let active = session.activeWorkout(in: context) else { return .clear }
        return .blocked(active: active, canDiscard: active.completedSetCount == 0)
    }

    @discardableResult
    static func start(_ request: Request, in context: ModelContext, session: WorkoutSessionController) -> Workout {
        switch request {
        case .routine(let routine): start(routine: routine, in: context, session: session)
        case .empty: startEmpty(in: context, session: session)
        }
    }

    /// "Discard it and start new": drops the workout in progress (only when nothing was completed in it) and starts
    /// `request`. With completed sets it refuses and brings the running workout back instead.
    /// The caller skips the rest timer first.
    @discardableResult
    static func discardAndStart(_ active: Workout, then request: Request,
                                in context: ModelContext, session: WorkoutSessionController) -> Workout {
        guard discardForNewStart(active, in: context, session: session) else { return active }
        return start(request, in: context, session: session)
    }

    /// The first half of "Discard it and start new": drops the workout in progress at once, so the mini bar goes
    /// with it, and returns true. With completed sets it refuses, brings the running workout back and returns false.
    static func discardForNewStart(_ active: Workout, in context: ModelContext, session: WorkoutSessionController) -> Bool {
        guard active.completedSetCount == 0 else {
            session.expand()
            return false
        }
        session.discard(active, context: context)
        return true
    }

    /// Starts a workout from a routine: one `WorkoutExercise` per item, `targetSets` rows each,
    /// prefilled from the last completed sets of that exercise (target reps when it was never done).
    @discardableResult
    static func start(routine: Routine, in context: ModelContext, session: WorkoutSessionController) -> Workout {
        let workout = Workout(name: routine.name)
        context.insert(workout)
        let defaultRest = defaultRestSeconds(in: context)
        var order = 0
        for item in routine.sortedItems {
            guard let exercise = item.exercise else { continue }
            let rest = item.restSeconds > 0
                ? item.restSeconds
                : RoutineSeeder.restSeconds(for: exercise.id, defaultRest: defaultRest)
            let added = append(exercise, to: workout, order: order, setCount: max(1, item.targetSets),
                               targetReps: item.targetReps, restSeconds: rest, in: context)
            added.supersetGroup = item.supersetGroup
            order += 1
        }
        // An item whose exercise was deleted may have split a superset.
        let started = workout.sortedExercises
        for (entry, group) in zip(started, Superset.normalized(started.map(\.supersetGroup))) { entry.supersetGroup = group }
        try? context.save()
        session.begin(workout)
        return workout
    }

    /// Starts an empty workout named "Workout".
    @discardableResult
    static func startEmpty(in context: ModelContext, session: WorkoutSessionController) -> Workout {
        let workout = Workout(name: String(localized: "workout.defaultName"))
        context.insert(workout)
        try? context.save()
        session.begin(workout)
        return workout
    }

    /// Appends one exercise (with prefilled rows) to an existing workout and marks it used.
    /// Rows copy the last session's sets row by row; an exercise never done gets `targetReps` and no weight.
    /// Rest defaults to the user's Rest length setting (heavy compounds a little longer).
    @discardableResult
    static func append(_ exercise: Exercise, to workout: Workout, order: Int, setCount: Int = 3, targetReps: Int = 0,
                       restSeconds: Int? = nil, in context: ModelContext) -> WorkoutExercise {
        let rest = restSeconds ?? RoutineSeeder.restSeconds(for: exercise.id, defaultRest: defaultRestSeconds(in: context))
        let workoutExercise = WorkoutExercise(order: order, exercise: exercise, restSeconds: rest)
        context.insert(workoutExercise)
        workoutExercise.workout = workout

        let template = lastCompletedSets(for: exercise, excluding: workout)
        for index in 0..<setCount {
            let source = index < template.count ? template[index] : template.last
            let set = SetEntry(order: index, kind: .normal, weightKg: source?.weightKg ?? 0, reps: source?.reps ?? targetReps)
            context.insert(set)
            set.workoutExercise = workoutExercise
        }
        exercise.lastUsedAt = .now
        return workoutExercise
    }

    /// The user's default rest (Settings > Rest timer > Rest length), 90 s before a profile exists.
    static func defaultRestSeconds(in context: ModelContext) -> Int {
        var descriptor = FetchDescriptor<UserProfile>()
        descriptor.fetchLimit = 1
        let value = (try? context.fetch(descriptor))?.first?.defaultRestSeconds ?? 0
        return value > 0 ? value : RoutineSeeder.defaultRestSeconds
    }

    /// Completed working sets from the most recent finished workout that did `exercise`, in row order
    /// (the same session the table's Previous column reads).
    static func lastCompletedSets(for exercise: Exercise, excluding current: Workout? = nil) -> [SetEntry] {
        let candidates = exercise.usages.filter { usage in
            guard let workout = usage.workout, workout.endedAt != nil else { return false }
            if let current, workout.persistentModelID == current.persistentModelID { return false }
            return usage.sets.contains { $0.isCompleted && $0.kind != .warmup && $0.reps > 0 }
        }
        guard let latest = candidates.max(by: { ($0.workout?.startedAt ?? .distantPast) < ($1.workout?.startedAt ?? .distantPast) }) else {
            return []
        }
        return latest.sortedSets.filter { $0.isCompleted && $0.kind != .warmup && $0.reps > 0 }
    }
}

File: android/app/src/main/java/app/notomorrow/feature/workout/WorkoutStarter.kt
package app.notomorrow.feature.workout

import app.notomorrow.data.dao.ExerciseDao
import app.notomorrow.data.dao.ProfileDao
import app.notomorrow.data.dao.RoutineDao
import app.notomorrow.data.dao.WorkoutDao
import app.notomorrow.data.entity.SetEntryEntity
import app.notomorrow.data.entity.WorkoutEntity
import app.notomorrow.data.entity.WorkoutExerciseEntity
import app.notomorrow.data.relation.CompletedSetRow
import app.notomorrow.di.AppContainer
import app.notomorrow.model.SetKind
import app.notomorrow.service.RoutineSeeder
import app.notomorrow.service.WorkoutSessionController
import java.util.UUID
import kotlin.math.max

/**
 * Builds `Workout` graphs (exercises + set rows) from a routine or from nothing, and hands them to
 * the session — 1:1 port of `NoTomorrow/Features/Workout/WorkoutStarter.swift`. The one start
 * path for Train and Today, so both build the same workout.
 *
 * SwiftData writes the graph through the object relationships; Room needs the parent row first, so
 * every insert goes through [WorkoutDao.insertWorkoutExerciseWithSets], which is `@Transaction`.
 *
 * The prefill rule lives in [templateSets] / [prefilledSets] as pure functions: rows are seeded
 * from the **last** completed working sets of that exercise, once the template runs out every
 * remaining row repeats its last entry, and an exercise never done gets the routine's target reps
 * at no weight.
 */
object WorkoutStarter {

    /** The four tables a start touches. */
    class Stores(
        val routineDao: RoutineDao,
        val workoutDao: WorkoutDao,
        val exerciseDao: ExerciseDao,
        val profileDao: ProfileDao,
    ) {
        companion object {
            fun of(container: AppContainer) = Stores(
                routineDao = container.db.routineDao(),
                workoutDao = container.db.workoutDao(),
                exerciseDao = container.db.exerciseDao(),
                profileDao = container.db.profileDao(),
            )
        }
    }

    /** What a Start tap asks for. [Empty.name] is `workout.defaultName`, resolved by the caller. */
    sealed interface Request {
        data class Routine(val routineId: String) : Request
        data class Empty(val name: String) : Request
    }

    /**
     * Whether a Start tap may go ahead. Only one workout runs at a time: while one is in progress
     * the caller offers Resume, and "Discard it and start new" only when that workout has no
     * completed sets.
     */
    sealed interface Gate {
        data object Clear : Gate
        data class Blocked(val active: WorkoutEntity, val canDiscard: Boolean) : Gate
    }

    suspend fun gate(workoutDao: WorkoutDao, session: WorkoutSessionController): Gate {
        val active = session.activeWorkout() ?: return Gate.Clear
        return Gate.Blocked(active, canDiscard = workoutDao.completedSetCount(active.id) == 0)
    }

    /** Starts [request] and opens it full screen. Returns the new workout id, `null` when the routine is gone. */
    suspend fun start(
        request: Request,
        stores: Stores,
        session: WorkoutSessionController,
        now: Long = System.currentTimeMillis(),
    ): String? = when (request) {
        is Request.Routine -> start(stores, session, request.routineId, now)
        is Request.Empty -> startEmpty(stores.workoutDao, session, request.name, now)
    }

    /**
     * "Discard it and start new": drops the workout in progress (only when nothing was completed in
     * it) and starts [request]. With completed sets it refuses and brings the running workout back
     * instead. The caller skips the rest timer first.
     */
    suspend fun discardAndStart(
        active: WorkoutEntity,
        request: Request,
        stores: Stores,
        session: WorkoutSessionController,
        now: Long = System.currentTimeMillis(),
    ): String? {
        if (stores.workoutDao.completedSetCount(active.id) != 0) {
            session.expand()
            return active.id
        }
        session.discard(active.id)
        return start(request, stores, session, now)
    }

    /**
     * `start(routine:in:session:)` — one `WorkoutExercise` per routine item, `targetSets` rows each,
     * prefilled from the last time the exercise was done (target reps when it never was). An item
     * whose rest is [RoutineSeeder.INHERIT_REST] takes the user's default rest. Supersets carry over.
     */
    suspend fun start(
        stores: Stores,
        session: WorkoutSessionController,
        routineId: String,
        now: Long = System.currentTimeMillis(),
    ): String? {
        val routine = stores.routineDao.routineWithItems(routineId) ?: return null
        val workout = WorkoutEntity(
            id = UUID.randomUUID().toString(),
            name = routine.routine.name,
            startedAt = now,
        )
        stores.workoutDao.insertWorkout(workout)
        val defaultRest = defaultRestSeconds(stores.profileDao)

        // `guard let exercise = item.exercise` — a routine item whose exercise was deleted is skipped,
        // which may split a superset: the groups are normalized over the items that stay.
        val kept = routine.sortedItems.filter { it.exercise != null }
        val groups = Superset.normalized(kept.map { it.item.supersetGroup })
        for ((order, entry) in kept.withIndex()) {
            val exerciseId = entry.exercise?.id ?: continue
            append(
                workoutDao = stores.workoutDao,
                exerciseDao = stores.exerciseDao,
                workoutId = workout.id,
                exerciseId = exerciseId,
                order = order,
                setCount = max(1, entry.item.targetSets),
                targetReps = entry.item.targetReps,
                restSeconds = routineItemRest(entry.item.restSeconds, exerciseId, defaultRest),
                defaultRest = defaultRest,
                supersetGroup = groups[order],
                now = now,
            )
        }

        session.begin(workout.id)
        return workout.id
    }

    /** `startEmpty(in:session:)`. [name] is `workout.defaultName` resolved by the caller. */
    suspend fun startEmpty(
        workoutDao: WorkoutDao,
        session: WorkoutSessionController,
        name: String,
        now: Long = System.currentTimeMillis(),
    ): String {
        val workout = WorkoutEntity(id = UUID.randomUUID().toString(), name = name, startedAt = now)
        workoutDao.insertWorkout(workout)
        session.begin(workout.id)
        return workout.id
    }

    /**
     * [append] against the container — the form the exercise picker uses. Rest defaults to the
     * user's Rest length setting (heavy compounds a little longer).
     */
    suspend fun append(
        container: AppContainer,
        workoutId: String,
        exerciseId: String,
        order: Int,
        setCount: Int = DEFAULT_SET_COUNT,
        restSeconds: Int? = null,
        now: Long = System.currentTimeMillis(),
    ): Long = append(
        workoutDao = container.db.workoutDao(),
        exerciseDao = container.db.exerciseDao(),
        workoutId = workoutId,
        exerciseId = exerciseId,
        order = order,
        setCount = setCount,
        restSeconds = restSeconds,
        defaultRest = defaultRestSeconds(container.db.profileDao()),
        now = now,
    )

    /**
     * `append(_:to:order:setCount:targetReps:restSeconds:in:)` — appends one exercise with
     * prefilled rows to an existing workout and stamps `lastUsedAt`. Returns the new
     * `workout_exercise` row id.
     */
    suspend fun append(
        workoutDao: WorkoutDao,
        exerciseDao: ExerciseDao,
        workoutId: String,
        exerciseId: String,
        order: Int,
        setCount: Int = DEFAULT_SET_COUNT,
        targetReps: Int = 0,
        restSeconds: Int? = null,
        defaultRest: Int = RoutineSeeder.DEFAULT_REST_SECONDS,
        supersetGroup: Int? = null,
        now: Long = System.currentTimeMillis(),
    ): Long {
        val rest = restSeconds ?: RoutineSeeder.restSeconds(exerciseId, defaultRest)
        val template = lastCompletedSets(workoutDao, exerciseId, excludingWorkoutId = workoutId)
        val id = workoutDao.insertWorkoutExerciseWithSets(
            WorkoutExerciseEntity(
                workoutId = workoutId,
                exerciseId = exerciseId,
                order = order,
                restSeconds = rest,
                supersetGroup = supersetGroup,
            ),
            prefilledSets(setCount, template, targetReps),
        )
        exerciseDao.markUsed(exerciseId, now)
        return id
    }

    /** The user's default rest (Settings > Rest timer > Rest length), 90 s before a profile exists. */
    suspend fun defaultRestSeconds(profileDao: ProfileDao): Int =
        resolveDefaultRest(profileDao.profile()?.defaultRestSeconds)

    /**
     * `lastCompletedSets(for:excluding:)` — the completed working sets of the most recent **other**
     * finished workout that contains the exercise, in row order (the same session the table's
     * Previous column reads).
     */
    suspend fun lastCompletedSets(
        workoutDao: WorkoutDao,
        exerciseId: String,
        excludingWorkoutId: String? = null,
    ): List<CompletedSetRow> =
        templateSets(workoutDao.completedSetsForExercise(exerciseId), excludingWorkoutId)

    // MARK: - Pure derivations

    /** A stored default of 0 or less (never written by Settings) falls back to 90 s. */
    fun resolveDefaultRest(stored: Int?): Int =
        stored?.takeIf { it > 0 } ?: RoutineSeeder.DEFAULT_REST_SECONDS

    /** A routine item's own rest, or — when it inherits (0) — the exercise's rest at [defaultRest]. */
    fun routineItemRest(itemRest: Int, exerciseId: String, defaultRest: Int): Int =
        if (itemRest > 0) itemRest else RoutineSeeder.restSeconds(exerciseId, defaultRest)

    /**
     * The prefill template: working sets (completed, not a warm-up, `reps > 0`) of the latest usage
     * of the exercise in a **finished** workout outside [excludingWorkoutId], sorted by row order.
     *
     * Ties on `startedAt` — the same workout holding the exercise twice — resolve to the later
     * `WorkoutExercise`. Swift's `Sequence.max(by:)` only replaces on a strict increase, so it keeps
     * the **first** maximum of the unordered `exercise.usages` set, i.e. an arbitrary one of the two;
     * `.thenBy { workoutExerciseOrder }` is a deliberate stabilisation of that, not a match.
     */
    fun templateSets(
        rows: List<CompletedSetRow>,
        excludingWorkoutId: String? = null,
    ): List<CompletedSetRow> {
        val working = rows.filter { row ->
            row.kind != SetKind.Warmup &&
                row.reps > 0 &&
                row.workoutEndedAt != null &&
                (excludingWorkoutId == null || row.workoutId != excludingWorkoutId)
        }
        val latest = working.maxWithOrNull(
            compareBy<CompletedSetRow> { it.workoutStartedAt }.thenBy { it.workoutExerciseOrder },
        ) ?: return emptyList()
        return working
            .filter { it.workoutExerciseId == latest.workoutExerciseId }
            .sortedBy { it.setOrder }
    }

    /**
     * `for index in 0..<setCount { source = index < template.count ? template[index] : template.last }`
     * — the rows a new exercise starts with; with no template, [targetReps] at no weight.
     * `workoutExerciseId` is filled in by the DAO.
     */
    fun prefilledSets(
        setCount: Int,
        template: List<CompletedSetRow>,
        targetReps: Int = 0,
    ): List<SetEntryEntity> =
        (0 until setCount).map { index ->
            val source = template.getOrNull(index) ?: template.lastOrNull()
            SetEntryEntity(
                workoutExerciseId = 0L,
                order = index,
                kind = SetKind.Normal,
                weightKg = source?.weightKg ?: 0.0,
                reps = source?.reps ?: targetReps,
            )
        }

    /** `setCount: Int = 3` on the iOS signature. */
    const val DEFAULT_SET_COUNT = 3
}
